compiler/noirc_frontend/src/hir/def_collector/dc_crate.rs

```diff
@@ -444,48 +465,57 @@ fn filter_literal_globals(
     })
 }
 
-fn type_check_globals(
-    interner: &mut NodeInterner,
-    global_ids: Vec<(FileId, GlobalId)>,
-) -> Vec<(CompilationError, fm::FileId)> {
-    global_ids
-        .into_iter()
-        .flat_map(|(file_id, global_id)| {
-            TypeChecker::check_global(global_id, interner)
-                .iter()
-                .cloned()
-                .map(|e| (e.into(), file_id))
-                .collect::<Vec<_>>()
-        })
-        .collect()
-}
+impl ResolvedModule {
+    fn type_check(&mut self, context: &mut Context) {
+        self.type_check_globals(&mut context.def_interner);
+        self.type_check_functions(&mut context.def_interner);
+        self.type_check_trait_impl_function(&mut context.def_interner);
+    }
 
-fn type_check_functions(
-    interner: &mut NodeInterner,
-    file_func_ids: Vec<(FileId, FuncId)>,
-) -> Vec<(CompilationError, fm::FileId)> {
-    file_func_ids
-        .into_iter()
-        .flat_map(|(file, func)| {
-            type_check_func(interner, func)
-                .into_iter()
-                .map(|e| (e.into(), file))
-                .collect::<Vec<_>>()
-        })
-        .collect()
-}
+    fn type_check_globals(&mut self, interner: &mut NodeInterner) {
+        for (file_id, global_id) in self.globals.iter() {
+            for error in TypeChecker::check_global(*global_id, interner) {
+                self.errors.push((error.into(), *file_id));
+            }
+        }
+    }
+
+    fn type_check_functions(&mut self, interner: &mut NodeInterner) {
+        for (file, func) in self.functions.iter() {
+            for error in type_check_func(interner, *func) {
+                self.errors.push((error.into(), *file));
+            }
+        }
+    }
+
+    fn type_check_trait_impl_function(&mut self, interner: &mut NodeInterner) {
+        for (file, func) in self.trait_impl_functions.iter() {
+            for error in check_trait_impl_method_matches_declaration(interner, *func) {
+                self.errors.push((error.into(), *file));
+            }
+            for error in type_check_func(interner, *func) {
+                self.errors.push((error.into(), *file));
+            }
+        }
+    }
+
+    /// Evaluate all `comptime` expressions in this module
+    fn evaluate_comptime(&self, interner: &mut NodeInterner) {
+        let mut interpreter = Interpreter::new(interner);
+
+        for (_file, function) in &self.functions {
+            interpreter.scan_function(*function);
```

Comment:
In the same vein as my other comment about the unwrap inside of `scan_function`, shouldn't we thread that error up and push onto `self.errors` here?

Reply:
We should add a compilation test that triggers an expected `InterpreterError`

Reply:
Yep, I'd just need a conversion function for the error but this PR was already getting too large